File: jdbc/src/test/java/net/snowflake/jdbc/e2e/put_get/PutGetAutoCompressTest.java
package net.snowflake.jdbc.e2e.put_get;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.nio.file.Files;
import java.nio.file.Path;
import java.sql.Connection;
import java.util.List;
import net.snowflake.jdbc.utils.SnowflakeIntegrationTestBase;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

public class PutGetAutoCompressTest extends SnowflakeIntegrationTestBase implements WithPutGet {

  private Path uncompressedFile() {
    return sharedTestDataDir().resolve("compression").resolve("test_data.csv");
  }

  @Test
  public void shouldCompressTheFileBeforeUploadingToStageWhenAutoCompressSetToTrue(
      @TempDir Path downloadDir) throws Exception {
    // Given Snowflake client is logged in
    Connection connection = getDefaultConnection();

    // When File is uploaded to stage with AUTO_COMPRESS set to true
    String stageName =
        createStageAndUploadFile(
            connection, "TEST_PUT_GET_AUTO_COMPRESS_TRUE", uncompressedFile(), true, true);

    // Then Only compressed file should be downloaded
    List<GetRow> rows =
        getFileFromStage(getDefaultConnection(), stageName, "test_data.csv", downloadDir);
    assertEquals(1, rows.size(), "Expected exactly one downloaded file");
    assertEquals("DOWNLOADED", rows.get(0).status, "Expected DOWNLOADED status");
    Path compressed = downloadDir.resolve("test_data.csv.gz");
    assertTrue(Files.exists(compressed), "Expected the compressed file on disk");
    assertFalse(
        Files.exists(downloadDir.resolve("test_data.csv")),
        "Did not expect an uncompressed file on disk");

    // And Have correct content
    assertEquals("1,2,3", readTextMaybeGzip(compressed).trim(), "Unexpected downloaded content");
  }

  @Test
  public void shouldNotCompressTheFileBeforeUploadingToStageWhenAutoCompressSetToFalse(
      @TempDir Path downloadDir) throws Exception {
    // Given Snowflake client is logged in
    Connection connection = getDefaultConnection();

    // When File is uploaded to stage with AUTO_COMPRESS set to false
    String stageName =
        createStageAndUploadFile(
            connection, "TEST_PUT_GET_AUTO_COMPRESS_FALSE", uncompressedFile(), false, true);

    // Then Only uncompressed file should be downloaded
    List<GetRow> rows =
        getFileFromStage(getDefaultConnection(), stageName, "test_data.csv", downloadDir);
    assertEquals(1, rows.size(), "Expected exactly one downloaded file");
    assertEquals("DOWNLOADED", rows.get(0).status, "Expected DOWNLOADED status");
    Path uncompressed = downloadDir.resolve("test_data.csv");
    assertTrue(Files.exists(uncompressed), "Expected the uncompressed file on disk");
    assertFalse(
        Files.exists(downloadDir.resolve("test_data.csv.gz")),
        "Did not expect a compressed file on disk");

    // And Have correct content
    assertEquals("1,2,3", readTextMaybeGzip(uncompressed).trim(), "Unexpected downloaded content");
  }
}

File: nodejs/tests/e2e/query/put-get/put-get-auto-compress.test.ts
import fs from 'node:fs';
import path from 'node:path';
import zlib from 'node:zlib';
import { describe, it, expect, beforeAll, afterAll, beforeEach, afterEach } from 'vitest';
import type { Connection } from '../../../types/sdk-types.js';
import { sharedTestDataDir, toFileUrl } from '../../utils/files.js';
import { createLiveConnection, createTempDir } from '../../utils/fixtures.js';
import { destroyConnectionAsync, executeAsync, randomizeName } from '../../utils/index.js';

describe('tests/definitions/shared/put_get/put_get_auto_compress.feature', () => {
  const TEST_FILE = path.join(sharedTestDataDir(), 'compression', 'test_data.csv');

  let connection: Connection;
  let stage: string;

  beforeAll(async () => {
    connection = await createLiveConnection({}, false);
  });

  afterAll(async () => {
    await destroyConnectionAsync(connection);
  });

  beforeEach(async () => {
    stage = randomizeName('TEST_PUT_GET_AUTO_COMPRESS');
    await executeAsync(connection, `CREATE TEMPORARY STAGE IF NOT EXISTS ${stage}`);
  });

  afterEach(async () => {
    await executeAsync(connection, `DROP STAGE IF EXISTS ${stage}`);
  });

  it('should compress the file before uploading to stage when AUTO_COMPRESS set to true', async () => {
    // Given Snowflake client is logged in
    void connection;

    // When File is uploaded to stage with AUTO_COMPRESS set to true
    await executeAsync(
      connection,
      `PUT ${toFileUrl(TEST_FILE)} @${stage} AUTO_COMPRESS=TRUE OVERWRITE=TRUE`,
    );

    const downloadDir = createTempDir();
    // Then Only compressed file should be downloaded
    const { rows } = await executeAsync(
      connection,
      `GET @${stage}/test_data.csv ${downloadDir.fileUrl()}/`,
    );
    expect(rows[0].status).toBe('DOWNLOADED');

    const downloaded = fs.readdirSync(downloadDir.path);
    expect(downloaded).toContain('test_data.csv.gz');
    expect(downloaded).not.toContain('test_data.csv');

    // And Have correct content
    const content = zlib.gunzipSync(fs.readFileSync(downloadDir.resolve('test_data.csv.gz')));
    expect(content.toString().trim()).toBe('1,2,3');
  });

  it('should not compress the file before uploading to stage when AUTO_COMPRESS set to false', async () => {
    // Given Snowflake client is logged in
    void connection;

    // When File is uploaded to stage with AUTO_COMPRESS set to false
    await executeAsync(
      connection,
      `PUT ${toFileUrl(TEST_FILE)} @${stage} AUTO_COMPRESS=FALSE OVERWRITE=TRUE`,
    );

    const downloadDir = createTempDir();
    // Then Only uncompressed file should be downloaded
    const { rows } = await executeAsync(
      connection,
      `GET @${stage}/test_data.csv ${downloadDir.fileUrl()}/`,
    );
    expect(rows[0].status).toBe('DOWNLOADED');

    const downloaded = fs.readdirSync(downloadDir.path);
    expect(downloaded).toContain('test_data.csv');
    expect(downloaded).not.toContain('test_data.csv.gz');

    // And Have correct content
    expect(fs.readFileSync(downloadDir.resolve('test_data.csv'), 'utf8').trim()).toBe('1,2,3');
  });
});
